milli/src/update/index_documents/extract/extract_geo_points.rs

```diff
@@ -24,19 +27,43 @@ pub fn extract_geo_points<R: io::Read + io::Seek>(
     let mut cursor = obkv_documents.into_cursor()?;
     while let Some((docid_bytes, value)) = cursor.move_on_next()? {
         let obkv = obkv::KvReader::new(value);
-        let (lat, lng) = obkv.get(lat_fid).zip(obkv.get(lng_fid)).ok_or_else(|| {
+        // since we only needs the primary key when we throw an error we create this getter to
+        // lazily get it when needed
+        let primary_key = || -> Value {
             let primary_key = obkv.get(primary_key_id).unwrap();
-            let primary_key = serde_json::from_slice(primary_key).unwrap();
-            UserError::InvalidGeoField { document_id: primary_key }
+            serde_json::from_slice(primary_key).unwrap()
+        };
+
+        // first we get the two fields
+        let lat = obkv.get(lat_fid).ok_or_else(|| -> UserError {
+            GeoError::MissingLatitude { document_id: primary_key() }.into()
+        })?;
+        let lng = obkv.get(lng_fid).ok_or_else(|| -> UserError {
+            GeoError::MissingLongitude { document_id: primary_key() }.into()
         })?;
-        let (lat, lng): (f64, f64) = (
-            serde_json::from_slice(lat).map_err(InternalError::SerdeJson)?,
-            serde_json::from_slice(lng).map_err(InternalError::SerdeJson)?,
-        );
+
+        // then we extract the values
+        let lat = extract_value(serde_json::from_slice(lat).map_err(InternalError::SerdeJson)?)
+            .map_err(|lat| -> UserError {
+                GeoError::BadLatitude { document_id: primary_key(), value: lat }.into()
+            })?;
+
+        let lng = extract_value(serde_json::from_slice(lng).map_err(InternalError::SerdeJson)?)
+            .map_err(|lng| -> UserError {
+                GeoError::BadLongitude { document_id: primary_key(), value: lng }.into()
+            })?;
 
         let bytes: [u8; 16] = concat_arrays![lat.to_ne_bytes(), lng.to_ne_bytes()];
         writer.insert(docid_bytes, bytes)?;
     }
 
     Ok(writer_into_reader(writer)?)
 }
+
+fn extract_value(value: Value) -> StdResult<f64, Value> {
```

Comment:
```suggestion
fn extract_float_from_value(value: Value) -> StdResult<f64, Value> {
```

Reply:
I did that + implemented `From<GeoError> for Error` like we did for the `FstError` and other things